kubeadm: add mutation for Linux paths in KubeletConfiguration on Windows

hwdef opened the pull request:
fix absolute paths do not work properly in config files in windows

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

<!--
Add one of the following kinds:

/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Because linux and windows express the absolute path differently, some errors may be caused.

This pr fixes this bug, the fix method comes from https://gist.github.com/neolit123/1b7375c8a956155a2ca1cdd901336bce


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes # https://github.com/kubernetes/kubeadm/issues/2419

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kubeadm: fix a bug on Windows worker nodes, where the downloaded KubeletConfiguration from the cluster can contain Linux paths that do not work on Windows and can trip the kubelet binary.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

hwdef committed:
kubeadm: fix absolute paths do not work properly in config files in windows

neolit123 commented:
thanks, i will have a look tomorrow.

neolit123 commented on cmd/kubeadm/app/componentconfigs/kubelet.go:
```diff
@@ -212,6 +216,63 @@ func (kc *kubeletConfig) Default(cfg *kubeadmapi.ClusterConfiguration, _ *kubead
+	}
+
+	// Mutate the fields we care about.
+	mutateStringField("resolvConf", cfg.ResolverConfig, true) // make sure this is disabled on Windows
```
i guess this was causing the panic you saw when testing?

neolit123 reviewed (commented):
/ok-to-test
/triage accepted
/priority backlog
/sig windows

neolit123 commented:
/release-note-edit

```release-note
kubeadm: fix a bug on Windows worker nodes, where the downloaded KubeletConfiguration from the cluster can contain Linux paths that do not work on Windows and can trip the kubelet binary.
```

neolit123 commented:
given i wrote this a while back in [gist](https://github.com/kubernetes/kubeadm/issues/2419) the change looks fine to me, but i would like to get a second opinion on the idea to have Mutate() on the component config.

/cc @SataQiu @pacoxu
any objections to that?

SataQiu commented on cmd/kubeadm/app/componentconfigs/kubelet.go:
```diff
@@ -212,6 +216,63 @@ func (kc *kubeletConfig) Default(cfg *kubeadmapi.ClusterConfiguration, _ *kubead
 
+// Mutate modifies absolute path fields in the KubeletConfiguration to be Windows compatible absolute paths.
+func (kc *kubeletConfig) Mutate() error {
+	if runtime.GOOS != "windows" {
```
It seems ok to add the `Mutate() error` function as a post hook before writing config to the disk. 
Not sure we should check the os by `runtime.GOOS != "windows"` or `// +build !windows`.

neolit123 replied:
> It seems ok to add the Mutate() error function as a post hook before writing config to the disk.

Mutate operates on the KubeletConfiguration object and Marshal finalizes the object converting it to yaml string, thus it has to be before Marshal.

> Not sure we should check the os by runtime.GOOS != "windows" or // +build !windows

+build !windows is better, but one problem is that if we move the *windows logic to _windows.go files, the unit test will not be run...currently the k8s CI only runs the Linux unit tests.

@hwdef i think we can have a TODO above the `if runtime.GOOS != "windows" line
> TODO: use build tags and move the Windows related logic to _windows.go files once the kubeadm code base is unit tested for Windows as part of CI - "GOOS=windows go test ...".

(please keep the commits squashed)

pacoxu commented:
In my mind, the path of config should be mutated in `default()` just after initial processing.

A general pattern is like
1. init()
2. default()
3. validate()
4. mutate()*
5. marshal() 

1-3 steps are getting the configuration.
With this `Mutate`, the pattern becomes a little complex.

I would like to add it to default() method if possible. However, the `Mutate` looks good. 
- Currently, `Mutate()` has only one scenario of `mutate paths` for Windows.

neolit123 commented:
> Currently, Mutate() has only one scenario of mutate paths for Windows.

yep, that is the only purpose. most importantly it's not part of publicly exposed function API and if the Windows mutation is one day no longer needed we can just remove it...

technically we cannot have this logic as part of Default, because the current path mutation is also updating potential user values for paths that the defaulting will not touch. e.g. 1) user adds a value that works on Linux but does not on Windows, 2) we mutate it, 3) we marshal.

neolit123 commented:
@hwdef please remove the `[WIP]` once you feel this is ready to merge.

hwdef commented:
> @hwdef please remove the `[WIP]` once you feel this is ready to merge.

@neolit123  I added TODO and deleted WIP.But I think we haven’t dealt with panic yet, do we have to deal with this problem?

pacoxu commented:
/retest
/lgtm

neolit123 commented:
> > @hwdef please remove the `[WIP]` once you feel this is ready to merge.
> 
> @neolit123 I added TODO and deleted WIP.But I think we haven’t dealt with panic yet, do we have to deal with this problem?

Hm, i though we deatlh with the panic by using poiner for resolverConf. Is there another panic?

To be safe we can always check if sub structures are not nil before apllying the mutations.

/approve
/hold

neolit123 commented on cmd/kubeadm/app/componentconfigs/kubelet.go:
```diff
@@ -212,6 +216,65 @@ func (kc *kubeletConfig) Default(cfg *kubeadmapi.ClusterConfiguration, _ *kubead
+	// Mutate the fields we care about.
+	mutateStringField("resolvConf", cfg.ResolverConfig, true) // make sure this is disabled on Windows
+	mutateStringField("staticPodPath", &cfg.StaticPodPath, false)
+	mutateStringField("authentication.x509.clientCAFile", &cfg.Authentication.X509.ClientCAFile, false)
```
This can be done only if the sub struct is not nil.

hwdef replied:
I updated the code, and do a test according to https://gist.github.com/neolit123/1b7375c8a956155a2ca1cdd901336bce#file-test-mutate-kubeletconfig-windows-md

I probably got the correct result
![image](https://user-images.githubusercontent.com/13084946/141042711-2aae722f-930a-4961-a8a2-7e1e85964bf8.png)

Please check whether the judgment code I added is correct, thank you!

neolit123 replied:
so there was a problem that we cannot leave the ResolverConfig field to `nil`. otherwise it will get defaulted to this on Windows https://github.com/kubernetes/kubernetes/blob/656d00e8949f64cd663ca532d12e00982a6664c4/pkg/kubelet/types/constants.go#L21 https://github.com/kubernetes/kubernetes/blob/ef0eff14ab41f43b9f5f314b20489eb74911d414/pkg/kubelet/apis/config/v1beta1/defaults.go#L185 and it would fail on runtime.

https://github.com/kubernetes/kubernetes/blob/33de444861d3de783a6618be9d10fa84da1c11b4/pkg/kubelet/kubelet.go#L1995

this still seems like a problem, so we should do this change in mutatePathsOnWindows:

```go

import (

utilpointer "k8s.io/utils/pointer"
...

)

func mutatePathsOnWindows(cfg *kubeletconfig.KubeletConfiguration, drive string) {
	mutateStringField := func(name string)
		// path.IsAbs() is not reliable here in the Windows runtime, so check if the
		// path starts with "/" instead. This means the path originated from a Unix node and
		// is an absolute path.
		if !strings.HasPrefix(*field, "/") {
			return
		}
		// Prepend the drive letter to the path and update the field.
		*field = filepath.Join(drive, *field)
		klog.V(2).Infof("[componentconfig] kubelet/Windows: adapted path for field %q to %q", name, *field)
	}

	// Mutate the fields we care about.
	klog.V(2).Infof("[componentconfig] kubelet/Windows: adapting field \"resolverConfig\" to empty")
	if cfg.ResolverConfig == nil {
		cfg.ResolverConfig = utilpointer.String("") // Make sure this is empty on Windows
	}
	mutateStringField("staticPodPath", &cfg.StaticPodPath)
	mutateStringField("authentication.x509.clientCAFile", &cfg.Authentication.X509.ClientCAFile)
}
```

+ might need test update.

RA489 commented:
/retest

hwdef commented on cmd/kubeadm/app/componentconfigs/kubelet.go:
```diff
@@ -17,8 +17,12 @@ limitations under the License.
+	"runtime"
+	"strings"
 
+	"github.com/pkg/errors"
```
If I add a new line after the pkg/errors import, the ci will report error.

neolit123 replied:
this works for other files.
https://github.com/kubernetes/kubernetes/blob/23065e23b54fe82aa7b42e04f5746f332e0e15b3/cmd/kubeadm/app/util/marshal.go#L24

what is the error?

neolit123 commented:
added a couple of new comments above:
https://github.com/kubernetes/kubernetes/pull/105992#discussion_r746766625
https://github.com/kubernetes/kubernetes/pull/105992#discussion_r746975330

note that code freeze for 1.23 is next week...otherwise the PR will have to wait for 1.24.

hwdef commented:
/test pull-kubernetes-unit

hwdef commented:
@neolit123 
I updated the code and the test code. And I did some test in windows node. It looks like normol.

And I add the blank line in `import`, I got the ci error again.I think these errors are weird. And I didn't find any solution, please take a look.

neolit123 commented:
the errors in CI are because there are multiple pkg/errors imports:
```
lubo@lubo-it:~/go/src/k8s.io/kubernetes$ git diff
diff --git a/cmd/kubeadm/app/componentconfigs/kubelet.go b/cmd/kubeadm/app/componentconfigs/kubelet.go
index 397965971cd..7f780dd6732 100644
--- a/cmd/kubeadm/app/componentconfigs/kubelet.go
+++ b/cmd/kubeadm/app/componentconfigs/kubelet.go
@@ -24,7 +24,6 @@ import (
 
        "github.com/pkg/errors"
 
-       "github.com/pkg/errors"
        "k8s.io/apimachinery/pkg/util/version"
        clientset "k8s.io/client-go/kubernetes"
        "k8s.io/klog/v2"
```

neolit123 commented:
also please add this missing unit test:
```
diff --git a/cmd/kubeadm/app/componentconfigs/kubelet_test.go b/cmd/kubeadm/app/componentconfigs/kubelet_test.go
index e1385bb67e4..09c9a0a62d6 100644
--- a/cmd/kubeadm/app/componentconfigs/kubelet_test.go
+++ b/cmd/kubeadm/app/componentconfigs/kubelet_test.go
@@ -350,6 +350,15 @@ func TestMutatePathsOnWindows(t *testing.T) {
                                },
                        },
                },
+               {
+                       name: "valid: resolverConfig is defaulted to empty string",
+                       cfg: &kubeletconfig.KubeletConfiguration{
+                               ResolverConfig: nil,
+                       },
+                       expected: &kubeletconfig.KubeletConfiguration{
+                               ResolverConfig: utilpointer.String(""),
+                       },
+               },
        }
 
        for _, test := range tests {
```

neolit123 commented:
/retitle kubeadm: add mutation for Linux paths in KubeletConfiguration on Windows

hwdef commented:
@neolit123  I updated the code and test code.
But regarding the ci error, it still has not been resolved.I checked the code and confirmed that only one `"github.com/pkg/errors"` package was imported. When I add a blank line after `"github.com/pkg/errors"`, ci will think that I have imported two error packages, After deleting this blank line, everything looks normal.

neolit123 commented:
i tested your patch locally and i saw 2 pkg error imports:

```
       "github.com/pkg/errors"
 
       "github.com/pkg/errors"
```

this is not visible in the github diff for some reason.

neolit123 commented:
the pkg/error import is already out of place in the current master branch:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/componentconfigs/kubelet.go#L22

so let's leave it like that...

neolit123 reviewed (commented):
thanks for the PR!
/lgtm

neolit123 commented:
/hold cancel
